pygments/lexers/python.py

```diff
@@ -1196,3 +1196,85 @@ def analyse_text(text):
         return (shebang_matches(text, r'pythonw?(3(\.\d)?)?') or
                 'import ' in ltext) \
             and ('import numpy' in ltext or 'from numpy import' in ltext)
+
+class XonshLexer(PythonLexer):
+    """
+    For Xonsh source code.
+    Based on internal XonshLexer: https://github.com/xonsh/xonsh/blob/main/xonsh/pyghooks.py
+    """
+
+    name = 'Xonsh'
+    url = 'https://xon.sh'
+    aliases = ['xonsh', 'xonshrc', 'xsh']
+    filenames = ['*.xsh', '*xonshrc']
+    mimetypes = ['text/x-xonsh', 'application/x-xonsh']
+    version_added = '2.19'
+
+    COMMAND_TOKEN_RE = r'[^=\s\[\]{}()$"\'`<&|;!]+(?=\s|$|\)|\]|\}|!)'
+
+    tokens = {
+        "mode_switch_brackets": [
+            (r"(\$)(\{)", bygroups(Keyword, Punctuation), "py_curly_bracket"),
+            (r"(@)(\()", bygroups(Keyword, Punctuation), "py_bracket"),
+            (
+                r"([\!\$])(\()",
+                bygroups(Keyword, Punctuation),
+                ("subproc_bracket", "subproc_start"),
+            ),
+            (
+                r"(@\$)(\()",
+                bygroups(Keyword, Punctuation),
+                ("subproc_bracket", "subproc_start"),
+            ),
+            (
+                r"([\!\$])(\[)",
+                bygroups(Keyword, Punctuation),
+                ("subproc_square_bracket", "subproc_start"),
+            ),
+            (r"(g?)(`)", bygroups(String.Affix, String.Backtick), "backtick_re"),
+        ],
+        "subproc_bracket": [(r"\)", Punctuation, "#pop"), include("subproc")],
+        "subproc_square_bracket": [(r"\]", Punctuation, "#pop"), include("subproc")],
+        "py_bracket": [(r"\)", Punctuation, "#pop"), include("root")],
+        "py_curly_bracket": [(r"\}", Punctuation, "#pop"), include("root")],
+        "backtick_re": [
+            (r"[\.\^\$\*\+\?\[\]\|]", String.Regex),
+            (r"({[0-9]+}|{[0-9]+,[0-9]+})\??", String.Regex),
+            (r"\\([0-9]+|[AbBdDsSwWZabfnrtuUvx\\])", String.Escape),
+            (r"`", String.Backtick, "#pop"),
+            (r"[^`\.\^\$\*\+\?\[\]\|]+", String.Backtick),
+        ],
+        "root": [
+            (r"\?", Keyword),
+            (r"(?<=\w)!", Keyword),
+            (r"\$\w+", Name.Variable),
+            (r"\(", Punctuation, "py_bracket"),
+            (r"\{", Punctuation, "py_curly_bracket"),
+            include("mode_switch_brackets"),
+            inherit,
+        ],
+        "subproc_start": [
+            (r"\s+", Whitespace),
+            (COMMAND_TOKEN_RE, Name.Builtin, "#pop"),
+            (r"", Whitespace, "#pop"),
```

Comment:
Pardon me, but I don't see this used anywhere else in Pygments, nor do I fully understand why you'd want to match nothing to pop? Can you elaborate?

Reply:
Comment from the sidelines: Use the `default("#pop")` idiom?

Reply:
Yes. I just realized it is on the "writing your lexer" page ... hmm.